### DISCUSSION TITLE
OpenAI account

### DISCUSSION THREAD

**Discussion by sarafarahabadi** [sovelluskontti/project-semantic-search#73] [General]
Hello Turo, 

I found a way to get the embeddings from OpenAI. I created a personal account to test the code but it does not have the sufficient usage to test the code. 
The code would be like below and I would have the OPEMAI_API_KEY stored in a env file. 
load_dotenv()
openai.api_key = os.getenv("OPENAI_API_KEY")

@app.route('/semantic-search', methods=['GET'])
def semantic_search():
    query = request.args.get('query', '')
    if not query:
        return jsonify({"error": "Query parameter is required"}), 400

    try:
        # Use OpenAI to generate the query embedding
        response = openai.Embedding.create(
            input=query,
            model="text-embedding-ada-002"
        )
        query_embedding = response['data'][0]['embedding']

        connection = get_db_connection()
        cursor = connection.cursor()
        try:
            # Perform the KNN search in Manticore
            sql = """
            SELECT id, title FROM movies 
            ORDER BY distance(embedding, query_embedding) ASC 
            LIMIT 10 OPTION query_embedding=%s;
            """
            cursor.execute(sql, (query_embedding,))
            results = cursor.fetchall()

            movies = [{"id": row[0], "title": row[1]} for row in results]
            return jsonify(movies)
        except Exception as e:
            return jsonify({"error": str(e)}), 500
        finally:
            cursor.close()
            connection.close()
    except Exception as e:
        return jsonify({"error": str(e)}), 500

Does Sovelluskontii oy has an OpenAI account with requires monthly payment so that it has sufficient usage limit to use for the project? 

If you have an account, please create a new API key for this project nd sent it to me so I can add it to the project and use it for getting the embeddings. Thank you so much. 

Best regards, 
Sara

**Comment by nyluntu** [comment]
Ok. I don't have an account at the moment. I will look at this how to proceed. Coming two days I don't have a much time because the Black Friday is coming and need to help my clients. 

So "fake it" is my hint for the moment. :) you can probably use some pre-generated embedded stuff to try searches that you have already done. Also there is probably something to read about the topics and get familiar with different technologies so this is good time for that if you need to wait my answer. I try to get back as soon as possible.

**Comment by nyluntu** [comment]
@sarafarahabadi 

I had time to check this. I created OpenAI project and invited you by your student email. When you have accepted the invitation I can add you to the project. I don't know if anything is really needed from the OpenAI platform site.

The API KEY for OpenAI is in shared folder. The text file contains link that shows you the api key. It should have required permissions.

You can try to use the api with the key. I have limited the budget be 10 dollars so when the limit is reached, the key will stop to work. Probably that is enough for testing anyway. When you use it, try to not commit the API key into github project. All secrets should not be part of the projects anyway. 

You probably have studying the API reference but here are some suggestions that I am interested about.

- `text-embedding-3-small` model should be working with this. It also be the lightest/cheapest option for embeddings.
- Try to learn how to use the embedding API to convert given input to embeddings. The length of the words/phrases are important when the costs are calculated. https://platform.openai.com/docs/api-reference/embeddings/create
- Try to understand the pricing policy and how it works. For programmers this is a good learning because many services have tricky pricing models. The OpenAI pricing is simple but not so easy to determine in advance.
- For embeddings, try the api if you can set how many dimensions are used. Default the model has like around 1500 dimensions and its probably too much and less works also.
- You probably need to regenerate the embeddings for the test data. There is a batch api for that kind of usage so look at this. https://platform.openai.com/docs/api-reference/batch/create
  - When you are familiar the embedding endpoint, use the batch endpoint to generate like 100 embeddings for the movies. We probably understand what the costs are and how much budget is used. If the embeddings are cheap to generate with batch endpoint, then we can decide how much more we need for testing. Also, save the batch endpoint results to shared files because there is no reason to generate same embeddings many times. 

This could be a good start for the coming week.
